src/compiler/checker.ts

```diff
@@ -12019,47 +11982,99 @@ namespace ts {
+            let stringIndexInfo: IndexInfo | undefined;
+            let applicableInfo: IndexInfo | undefined;
+            let applicableInfos: IndexInfo[] | undefined;
+            for (const info of indexInfos) {
+                if (info.keyType === stringType) {
+                    stringIndexInfo = info;
+                }
+                else if (isApplicableIndexType(keyType, info.keyType)) {
+                    if (!applicableInfo) {
+                        applicableInfo = info;
+                    }
+                    else {
+                        (applicableInfos || (applicableInfos = [applicableInfo])).push(info);
+                    }
+                }
+            }
+            // When more than one index signature is applicable we create a synthetic IndexInfo. Instead of computing
+            // the intersected key type, we just use unknownType for the key type as nothing actually depends on the
+            // keyType property of the returned IndexInfo.
+            return applicableInfos ? createIndexInfo(unknownType, getIntersectionType(map(applicableInfos, info => info.type)),
+                    reduceLeft(applicableInfos, (isReadonly, info) => isReadonly && info.isReadonly, /*initial*/ true)) :
+                applicableInfo ? applicableInfo :
+                stringIndexInfo && isApplicableIndexType(keyType, stringType) ? stringIndexInfo :
+                undefined;
+        }
+
+        function findApplicableIndexInfoForName(indexInfos: readonly IndexInfo[], name: __String) {
+            // We optimize for cases where we have no template literal index signatures as it is somewhat expensive
+            // to obtain the literal type for a name and check for assignability.
+            let stringIndexInfo: IndexInfo | undefined;
+            for (const info of indexInfos) {
+                const keyType = info.keyType;
+                if (keyType === stringType) {
+                    stringIndexInfo = info;
+                }
+                else if (keyType === numberType && isNumericLiteralName(name)) {
+                    return info;
+                }
+                else if (keyType.flags & TypeFlags.TemplateLiteral) {
+                    return findApplicableIndexInfo(indexInfos, getLiteralType(unescapeLeadingUnderscores(name)));
+                }
+            }
+            return stringIndexInfo;
+        }
+
+        function isApplicableIndexType(source: Type, target: Type) {
+            // A 'string' index signature applies to types assignable to 'string' or 'number', and a 'number' index
+            // signature applies to types assignable to 'number' and numeric string literal types.
+            return isTypeAssignableTo(source, target) ||
+                target === stringType && isTypeAssignableTo(source, numberType) ||
+                target === numberType && source.flags & TypeFlags.StringLiteral && isNumericLiteralName((source as StringLiteralType).value);

```

Comment:
There's also the type
```ts
type NumericStrings = `${number}`
```
which might warrant special-casing here, as well? It's a bit awkward because `"0.0"` is a "numeric string" but is actually a distinct object key string from `"0"`; but that's already true for the literals themselves, so.... ¯\\_(ツ)_/¯

Reply:
I'm not sure there's any meaningful special casing we could do for `` `${number}` `` because, as you already point out, it may or may not contain a round-trippable numeric string.